### PR TITLE
Multi note

### PR DESCRIPTION
**Pull request by 70ray**
This enables multi-line proofers' notes and allows nested pairs of brackets inside proofers' notes. Tasks https://www.pgdp.net/c/tasks.php?action=show&task_id=2001 and https://www.pgdp.net/c/tasks.php?action=show&task_id=1977.

Sandbox at: https://www.pgdp.org/~rp31/c.branch/multi_note

### CODE REVIEW

File: tools/proofers/preview.js [hunk 1]
```diff
@@ -26,16 +26,26 @@ function processExMath(text, textFunction, allowMath) {
...
-    function removeComments(textLine) {
-        return textLine.replace(/\[\*\*[^\]]*\]/g, '');
-    }
-
-    // true if txtLine contains anything except only comments or spaces
-    function nonComment(textLine) {
-        return (/\S/.test(removeComments(textLine)));
+// find index of next unmatched ] return 0 if none found
+function findClose(txt, index) {
+    var result;
```
Comment (chrismiceli) [inline tools/proofers/preview.js:31]: Should we use let and const like our other js files?

File: tools/proofers/preview.js [hunk 2]
```diff
@@ -26,16 +26,26 @@ function processExMath(text, textFunction, allowMath) {
...
+    var nestLevel = 0;
+    var re = /\[|\]/g;  // [ or ]
```
Comment (chrismiceli) [inline tools/proofers/preview.js:33]: Not sure if it matters, but I'd put the regex out of the function, so it won't be created every function invocation. Compiling regex can be expensive and I've seen this trick cause an app go from 8 min to instant.

File: tools/proofers/preview.js [hunk 3]
```diff
@@ -107,6 +117,65 @@ $(function () {
...
+            // find start of note
+            while (null != (result = reNoteStart.exec(txt))) {
+                noteStartIndex = result.index;
+                noteEndIndex = findClose(txt, reNoteStart.lastIndex);
+                // copy text to start of note
+                txtOut += txt.slice(beginIndex, noteStartIndex);
+                outNoteStart += noteStartIndex - beginIndex;
+                // if note starts at beginning of a line and ends at the end of a line
+                // remove any following nl character also so doesn't count as a blank line and include the nl in the note
+                if(((noteStartIndex === 0) || (txt.charAt(noteStartIndex - 1) === "\n")) && ((noteEndIndex === maxIndex) || (txt.charAt(noteEndIndex + 1) === "\n"))) {
```
Comment (chrismiceli) [inline tools/proofers/preview.js:160]: maybe add some newlines to group the conditions more clearly:
```
if(((noteStartIndex === 0) || (txt.charAt(noteStartIndex - 1) === "\n")) &&
   ((noteEndIndex === maxIndex) || (txt.charAt(noteEndIndex + 1) === "\n"))) {
```

File: tools/proofers/preview.js [hunk 4]
```diff
@@ -1125,6 +1065,57 @@ $(function () {
...
+        let issueStarts = [];
+        let issueEnds = [];
+        let errorString;
+        issArray.forEach(function(issue) {
+            if (issue.type === 0) {
+                errorString = makeErrStr("hlt");
+            } else {
+                errorString = makeErrStr("err");
+            }
+            let message = previewMessages[issue.code].replace("%s", issue.subText);
```
Comment (chrismiceli) [inline tools/proofers/preview.js:1079]: do we need to make sure issue.subText doesn't have any of the patterns used by replace (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_a_parameter). We ran into an issue like this in the find replace dialog, so you can see my solution there if necessary https://github.com/DistributedProofreaders/dproofreaders/blob/master/tools/proofers/srchrep.js#L27

Comment (srjfoo) [reply tools/proofers/preview.js:1079]: May not be an adequate test, but I used 

```
[** [386]][** [** [386]]$1\n\n\\n\\n$$2][** A][** B]
```
as proofers' notes, and what was shown in Preview was

```
[** B][** A][** [** [386]]$1\n\n\n\n$$2][** [386]]
```

I see two issues immediately 
* while the `$$` was accurately displayed in the preview, the `\\` was not. A single backslash appears to be echoed just fine, but if it's being used preceding another backslash or a `$`, it's treated as an escape character.
* There are four consecutive notes there -- they're displayed in the preview in reverse order.

File: tools/proofers/preview.js (hunk 1 above)
Comment (70ray) [reply tools/proofers/preview.js:31]: Yes.

File: tools/proofers/preview.js (hunk 2 above)
Comment (70ray) [reply tools/proofers/preview.js:33]: Interesting. Done as you suggest.

File: tools/proofers/preview.js (hunk 3 above)
Comment (70ray) [reply tools/proofers/preview.js:160]: good idea.

File: tools/proofers/preview.js (hunk 4 above)
Comment (70ray) [reply tools/proofers/preview.js:1079]: @chrismiceli The string is always predefined so it shouldn't be a problem. i.e. '/*', '/#', '*/', '#/', 'Footnote', 'Illustration', 'Sidenote', '&lt;tb&gt;'.

@srjfoo Interesting. I'll have to look into this.

Comment (70ray) [reply tools/proofers/preview.js:1075]: @srjfoo I've revised the sorting code so the notes appear in the correct order.
It looks like the backslash issue is caused by MathJax. If I disable the Math preview the backslashes in your example appear correctly. Does it work like that for you?

Comment (srjfoo) [reply tools/proofers/preview.js:1075]: Yep, the sorting looks good now.

And yes, turning off Math preview takes care of the backslash issue -- I didn't think of that. Is it fair to assume that that's something that will have to be lived with? (Bear in mind that I was using test strings that are unlikely to be used in reality to test per Chris' question).

Would it be appropriate to move the Math checkbox out of the "Configure" block and into the main Preview Toolbox? I realize we need to keep that small, but I had consciously decided "Well, there's no harm in leaving the Math checkbox checked permanently" and gone off and forgotten about it. If the checkbox is in the main toolbar, it's a reminder that "this is something you may want immediate on-off access to".

Comment (70ray) [reply tools/proofers/preview.js:1075]: > Yep, the sorting looks good now.
> 
> And yes, turning off Math preview takes care of the backslash issue -- I didn't think of that. Is it fair to assume that that's something that will have to be lived with? (Bear in mind that I was using test strings that are unlikely to be used in reality to test per Chris' question).
> 
> Would it be appropriate to move the Math checkbox out of the "Configure" block and into the main Preview Toolbox? I realize we need to keep that small, but I had consciously decided "Well, there's no harm in leaving the Math checkbox checked permanently" and gone off and forgotten about it. If the checkbox is in the main toolbar, it's a reminder that "this is something you may want immediate on-off access to".

I don't see why MathJax should need to change the backslash characters in this way. I'll enquire why it happens.
In the meantime we can live with it as we have been without noticing the problem. Moving the checkbox is something we could consider.

**Review by srjfoo — APPROVED** [review]
This looks to me like it solves the problems mentioned in both tasks. I'm happy to put off dealing with the disappearing-double-backslashes-when-math-mode-is-turned-on, and considering moving the checkbox to enable math mode off to a future MR (or to just footnote it as a hopefully rare side-effect of enabling math mode, if that's the better option).

I would recommend giving the formatting community a week or two to review it before merging (that still gives us time to get it in for December's code deployment, and US Thanksgiving week is next week).

**Comment by cpeel** [comment]
@70ray - is this ready for final testing and review?

@chrismiceli - can you please re-review when you have a moment?

**Comment by srjfoo** [comment]
I'm not sure whether or not this issue was part of the original MR, because I didn't think to test `<tb>`s at that time.

In the sandbox, with color markup turned on:
image

and turned off:
image

On both PROD and the main TEST sandbox, the `<tb>` tag stays there either way.

**Comment by 70ray** [comment]
> @70ray - is this ready for final testing and review?
> 
No there is still an issue with the <tb> tag as @srjfoo found.

**Comment by 70ray** [comment]
It's still not quite right.

**Comment by 70ray** [comment]
This should be ready for final testing and review now.

**Review by srjfoo — APPROVED** [review]
> This should be ready for final testing and review now.

I've re-tested, I think, and also did some spot checking in BrowserStack. Looks good to me.

I've also requested additional testing in the [New version of Format Preview available for testing.](https://www.pgdp.net/phpBB3/viewtopic.php?t=75636) topic.

**Review by chrismiceli — APPROVED** [review]

File: tools/proofers/preview.js [hunk 5]
```diff
@@ -26,16 +26,27 @@ function processExMath(text, textFunction, allowMath) {
...
-    function removeComments(textLine) {
-        return textLine.replace(/\[\*\*[^\]]*\]/g, '');
-    }
 
-    // true if txtLine contains anything except only comments or spaces
-    function nonComment(textLine) {
-        return (/\S/.test(removeComments(textLine)));
+// find index of next unmatched ] return 0 if none found
+const re = /\[|\]/g;  // [ or ]
+function findClose(txt, index) {
```
[the remark this answers is not in the document]
Comment (70ray) [reply tools/proofers/preview.js:46]: It is also tested in formatPreviewTests.js. Not sure if this is a good reason.

File: tools/proofers/preview.js [hunk 6]
```diff
@@ -92,21 +103,75 @@ $(function () {
...
+            txtOut += txt.slice(beginIndex);
+
+            // remove trailing whitespace on each line and trailing blank lines
+            txtOut = txtOut.replace(/ *$/mg, "");
+            return txtOut.replace(/\s*$/, "");
+        }
+
         // find end of line (or eot) following ix
         function findEnd(ix) {
             var re = /\n|$/g;
```
[the remark this answers is not in the document]
Comment (70ray) [reply tools/proofers/preview.js:201]: Yes, perhaps later.

**Comment by 70ray** [comment]
Davide Cervone of MathJax has replied to my query about the backslashes: "In order to allow you to include a literal $, \(, \), \[ or \] in your text, MathJax processes double backslashes as single backslashes.  So you can use \\( to get a literal \( in your page.  There is a configuration parameter that controls whether that processing is done or not, so you can turn that off if you wish."

**Comment by srjfoo** [comment]
> Davide Cervone of MathJax has replied to my query about the backslashes: "In order to allow you to include a literal $, (, ), [ or ] in your text, MathJax processes double backslashes as single backslashes. So you can use \( to get a literal ( in your page. There is a configuration parameter that controls whether that processing is done or not, so you can turn that off if you wish."

So, I guess the first question is, what do we want Preview to show? I guess at least for the "Tags" version, I expect everything to show in Preview. The thing is, "No Tags" really isn't fully "No Tags"; it's more properly "No inline tags", because all of the block level tags remain.

The second question that comes to mind is, can we proceed with this the way it is, with the caveat that backslashes have a special meaning in certain cases in MathJax, and that for now, if Math is turned on, backslashes used to escape special characters in the math markup will not be shown?

It's been a little over a week. I assume at least someone took the time to re-test, but we've heard 🦗 🦗 🦗 from the intended audience.

If there's some additional testing you'd like me to do, let me know.

**Comment by 70ray** [comment]
> > Davide Cervone of MathJax has replied to my query about the backslashes: "In order to allow you to include a literal $, (, ), [ or ] in your text, MathJax processes double backslashes as single backslashes. So you can use ( to get a literal ( in your page. There is a configuration parameter that controls whether that processing is done or not, so you can turn that off if you wish."
> 
> So, I guess the first question is, what do we want Preview to show? I guess at least for the "Tags" version, I expect everything to show in Preview. The thing is, "No Tags" really isn't fully "No Tags"; it's more properly "No inline tags", because all of the block level tags remain.
> 
Yes, "tags" means in-line tags in this context. `<tb>` although it looks like an in-line tag, is treated like an out-of-line tag for this purpose. We could make a new task to clarify this issue. 

> The second question that comes to mind is, can we proceed with this the way it is, with the caveat that backslashes have a special meaning in certain cases in MathJax, and that for now, if Math is turned on, backslashes used to escape special characters in the math markup will not be shown?
> 
I think that we can leave it like it is. I presume backslashes are unlikely to appear in any DP books. The single $ sign is by default not enabled as a delimiter for math so $ can appear without any problem. `\$` will also just produce a $ sign.

> It's been a little over a week. I assume at least someone took the time to re-test, but we've heard cricket cricket cricket from the intended audience.
> 
> If there's some additional testing you'd like me to do, let me know.

It is difficult to say. There always seems to be something that I hadn't thought of that catches us out. I came across one thing that could be considered a bug while testing but it is not new so not really in the scope of this.

**Comment by srjfoo** [comment]
> > It's been a little over a week. I assume at least someone took the time to re-test, but we've heard cricket cricket cricket from the intended audience.
> > If there's some additional testing you'd like me to do, let me know.
> 
> It is difficult to say. There always seems to be something that I hadn't thought of that catches us out. I came across one thing that could be considered a bug while testing but it is not new so not really in the scope of this.

I think we want to keep the scope focused on the two tasks that you addressed in this MR, plus any associated problems we uncovered along the way that are deemed critical to their operation, so I'm in favor of proceeding, given that we've had no complaints to do with it in over a week.

My collection of tests is on [this page in your sandbox](https://www.pgdp.org/~rp31/c.branch/multi_note/tools/page_browser.php?project=projectID44dda6ce1d66d&imagefile=141.png&mode=imageText&round_id=F1), if you want to take a look at it and see if you think there's anything I haven't covered. It passes in your sandbox, but fails on the main test site (yes, it contains multiple tests, which, for troubleshooting errors is not good, but when the problems are fixed, it's convenient to have them all in one place).

**Comment by 70ray** [comment]
> 
> My collection of tests is on [this page in your sandbox](https://www.pgdp.org/~rp31/c.branch/multi_note/tools/page_browser.php?project=projectID44dda6ce1d66d&imagefile=141.png&mode=imageText&round_id=F1), if you want to take a look at it and see if you think there's anything I haven't covered. It passes in your sandbox, but fails on the main test site (yes, it contains multiple tests, which, for troubleshooting errors is not good, but when the problems are fixed, it's convenient to have them all in one place).

It looks like a good test to me.

**Review by chrismiceli — APPROVED** [review]

**Pull request merged**